From f3eb4c81fa5a0bd51ea7ffb6fe8782af59c528c5 Mon Sep 17 00:00:00 2001
From: Jussi Laako <jussi@sonarnerd.net>
Date: Mon, 6 Jan 2020 20:41:42 +0200
Subject: [PATCH 5/9] Change DSD silence pattern to avoid clicks/pops

---
 sound/usb/endpoint.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/sound/usb/endpoint.c b/sound/usb/endpoint.c
index 8f65349a06d3..f729a2844182 100644
--- a/sound/usb/endpoint.c
+++ b/sound/usb/endpoint.c
@@ -1124,7 +1124,7 @@ static int data_ep_set_params(struct snd_usb_endpoint *ep)
 	case SNDRV_PCM_FORMAT_DSD_U32_LE:
 	case SNDRV_PCM_FORMAT_DSD_U16_BE:
 	case SNDRV_PCM_FORMAT_DSD_U32_BE:
-		ep->silence_value = 0x69;
+		ep->silence_value = 0xaa;
 		break;
 	default:
 		ep->silence_value = 0;
-- 
2.34.1

