From e9077dfbde8a70b7e36f6fc58136f8e585c72dcb Mon Sep 17 00:00:00 2001
From: Jussi Laako <jussi@sonarnerd.net>
Date: Fri, 22 Apr 2022 00:06:01 +0300
Subject: [PATCH 1/2] Add Primare devices to DSD auto-detect vendor list

---
 sound/usb/quirks.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/sound/usb/quirks.c b/sound/usb/quirks.c
index 300826791843..6570358decba 100644
--- a/sound/usb/quirks.c
+++ b/sound/usb/quirks.c
@@ -1753,6 +1753,7 @@ u64 snd_usb_interface_dsd_format_quirks(struct snd_usb_audio *chip,
 	case 0x21b4:  /* Wavelength Audio */
 	case 0x22d9:  /* Oppo */
 	case 0x23ba:  /* Playback Designs */
+	case 0x24a4:  /* Primare */
 	case 0x25ce:  /* Mytek devices */
 	case 0x278b:  /* Rotel? */
 	case 0x292b:  /* Gustard/Ess based devices */
-- 
2.34.1

