From ee57bed5e117b1960a3af4c19accfab3c788b389 Mon Sep 17 00:00:00 2001
From: Jussi Laako <jussi@sonarnerd.net>
Date: Mon, 28 Nov 2022 19:41:02 +0200
Subject: [PATCH] Add PS Audio and Comtrue to the DSD auto-detection list

---
 sound/usb/quirks.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/sound/usb/quirks.c b/sound/usb/quirks.c
index d538cf9755d0..316b18496a25 100644
--- a/sound/usb/quirks.c
+++ b/sound/usb/quirks.c
@@ -1756,11 +1756,13 @@ u64 snd_usb_interface_dsd_format_quirks(struct snd_usb_audio *chip,
 	case 0x23ba:  /* Playback Designs */
 	case 0x24a4:  /* Primare */
 	case 0x25ce:  /* Mytek devices */
+	case 0x2616:  /* PS Audio */
 	case 0x278b:  /* Rotel? */
 	case 0x292b:  /* Gustard/Ess based devices */
 	case 0x2972:  /* FiiO devices */
 	case 0x2ab6:  /* T+A devices */
 	case 0x2d87:  /* Cayin? */
+	case 0x2fc6:  /* Comtrue */
 	case 0x3353:  /* Khadas devices */
 	case 0x3842:  /* EVGA */
 	case 0xc502:  /* HiBy devices */
-- 
2.34.1

