diff -ur a/drivers/usb/core/config.c b/drivers/usb/core/config.c
--- a/drivers/usb/core/config.c	2022-10-03 05:09:07.000000000 +0800
+++ b/drivers/usb/core/config.c	2022-11-05 19:05:59.032455460 +0800
@@ -190,7 +190,7 @@
 };
 static const unsigned short high_speed_maxpacket_maxes[4] = {
 	[USB_ENDPOINT_XFER_CONTROL] = 64,
-	[USB_ENDPOINT_XFER_ISOC] = 1024,
+	[USB_ENDPOINT_XFER_ISOC] = 1544,
 
 	/* Bulk should be 512, but some devices use 1024: we will warn below */
 	[USB_ENDPOINT_XFER_BULK] = 1024,
@@ -198,7 +198,7 @@
 };
 static const unsigned short super_speed_maxpacket_maxes[4] = {
 	[USB_ENDPOINT_XFER_CONTROL] = 512,
-	[USB_ENDPOINT_XFER_ISOC] = 1024,
+	[USB_ENDPOINT_XFER_ISOC] = 1544,
 	[USB_ENDPOINT_XFER_BULK] = 1024,
 	[USB_ENDPOINT_XFER_INT] = 1024,
 };
