From 48b71564ad7680cee749ba2d7f17b0fad8e4379f Mon Sep 17 00:00:00 2001
From: Jussi Laako <jussi@sonarnerd.net>
Date: Wed, 16 Jan 2019 14:50:26 +0200
Subject: [PATCH 6/7] Adjust USB isochronous packet size

---
 drivers/usb/core/config.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/usb/core/config.c b/drivers/usb/core/config.c
index 00e28456e4cc..4bc8d05345ec 100644
--- a/drivers/usb/core/config.c
+++ b/drivers/usb/core/config.c
@@ -190,7 +190,7 @@ static const unsigned short full_speed_maxpacket_maxes[4] = {
 };
 static const unsigned short high_speed_maxpacket_maxes[4] = {
 	[USB_ENDPOINT_XFER_CONTROL] = 64,
-	[USB_ENDPOINT_XFER_ISOC] = 1024,
+	[USB_ENDPOINT_XFER_ISOC] = 1544,
 
 	/* Bulk should be 512, but some devices use 1024: we will warn below */
 	[USB_ENDPOINT_XFER_BULK] = 1024,
@@ -198,7 +198,7 @@ static const unsigned short high_speed_maxpacket_maxes[4] = {
 };
 static const unsigned short super_speed_maxpacket_maxes[4] = {
 	[USB_ENDPOINT_XFER_CONTROL] = 512,
-	[USB_ENDPOINT_XFER_ISOC] = 1024,
+	[USB_ENDPOINT_XFER_ISOC] = 1544,
 	[USB_ENDPOINT_XFER_BULK] = 1024,
 	[USB_ENDPOINT_XFER_INT] = 1024,
 };
-- 
2.34.1

