diff -ur a/sound/usb/endpoint.c b/sound/usb/endpoint.c
--- a/sound/usb/endpoint.c	2022-11-05 18:06:18.143864685 +0800
+++ b/sound/usb/endpoint.c	2022-11-05 18:25:01.113120479 +0800
@@ -1096,7 +1096,7 @@
 	case SNDRV_PCM_FORMAT_DSD_U32_LE:
 	case SNDRV_PCM_FORMAT_DSD_U16_BE:
 	case SNDRV_PCM_FORMAT_DSD_U32_BE:
-		ep->silence_value = 0x69;
+		ep->silence_value = 0xaa;
 		break;
 	default:
 		ep->silence_value = 0;
