From 2d7e073f900a6e241e209d2dad5f0b01002efd1d Mon Sep 17 00:00:00 2001
From: Emilio Moretti <emilio.moretti@gmail.com>
Date: Tue, 19 Nov 2019 06:08:48 -0300
Subject: [PATCH 05/12] Add VID to support native DSD reproduction on FiiO
 devices.

Signed-off-by: Emilio Moretti <emilio.moretti@gmail.com>
---
 sound/usb/quirks.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/sound/usb/quirks.c b/sound/usb/quirks.c
index 042aefc15e8a..7d10f59ac578 100644
--- a/sound/usb/quirks.c
+++ b/sound/usb/quirks.c
@@ -1715,6 +1715,7 @@ u64 snd_usb_interface_dsd_format_quirks(struct snd_usb_audio *chip,
 	case 0x25ce:  /* Mytek devices */
 	case 0x278b:  /* Rotel? */
 	case 0x292b:  /* Gustard/Ess based devices */
+	case 0x2972:  /* FiiO devices */
 	case 0x2ab6:  /* T+A devices */
 	case 0x3842:  /* EVGA */
 	case 0xc502:  /* HiBy devices */
-- 
2.17.1

