From 7544b043b52b2f56dfb2e41a4d8823694ebda2c7 Mon Sep 17 00:00:00 2001
From: Jussi Laako <jussi@sonarnerd.net>
Date: Mon, 6 Jan 2020 20:41:42 +0200
Subject: [PATCH 7/8] Change DSD silence pattern to avoid clicks/pops

---
 sound/usb/endpoint.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/sound/usb/endpoint.c b/sound/usb/endpoint.c
index 3bbc227769d0..39b9bba4c0a5 100644
--- a/sound/usb/endpoint.c
+++ b/sound/usb/endpoint.c
@@ -1054,7 +1054,7 @@ static int data_ep_set_params(struct snd_usb_endpoint *ep)
 	case SNDRV_PCM_FORMAT_DSD_U32_LE:
 	case SNDRV_PCM_FORMAT_DSD_U16_BE:
 	case SNDRV_PCM_FORMAT_DSD_U32_BE:
-		ep->silence_value = 0x69;
+		ep->silence_value = 0xaa;
 		break;
 	default:
 		ep->silence_value = 0;
-- 
2.34.1

