From 9a67710027ff742f3dfd21e32318d056b0c490f3 Mon Sep 17 00:00:00 2001
From: Alexandre Frade <kernel@xanmod.org>
Date: Sun, 29 May 2022 00:57:40 +0000
Subject: [PATCH 19/20] XANMOD: scripts/setlocalversion: remove '+' tag for git
 repo short version

Signed-off-by: Alexandre Frade <kernel@xanmod.org>
---
 scripts/setlocalversion | 1 -
 1 file changed, 1 deletion(-)

diff --git a/scripts/setlocalversion b/scripts/setlocalversion
index 28169d7e143b..6ce26e12342f 100755
--- a/scripts/setlocalversion
+++ b/scripts/setlocalversion
@@ -117,7 +117,6 @@ scm_version()
 		# If only the short version is requested, don't bother
 		# running further git commands
 		if $short; then
-			echo "+"
 			return
 		fi
 
-- 
2.47.2

