From 96d1790ae8a92213ab3063a0242b2be24511ec78 Mon Sep 17 00:00:00 2001
From: Jussi Laako <jussi@sonarnerd.net>
Date: Wed, 16 Jan 2019 14:50:26 +0200
Subject: [PATCH 4/9] Adjust USB isochronous packet size

---
 drivers/usb/core/config.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/usb/core/config.c b/drivers/usb/core/config.c
index 7f8d33f92ddb..cbfd75206544 100644
--- a/drivers/usb/core/config.c
+++ b/drivers/usb/core/config.c
@@ -190,7 +190,7 @@ static const unsigned short full_speed_maxpacket_maxes[4] = {
 };
 static const unsigned short high_speed_maxpacket_maxes[4] = {
 	[USB_ENDPOINT_XFER_CONTROL] = 64,
-	[USB_ENDPOINT_XFER_ISOC] = 1024,
+	[USB_ENDPOINT_XFER_ISOC] = 1544,
 
 	/* Bulk should be 512, but some devices use 1024: we will warn below */
 	[USB_ENDPOINT_XFER_BULK] = 1024,
@@ -198,7 +198,7 @@ static const unsigned short high_speed_maxpacket_maxes[4] = {
 };
 static const unsigned short super_speed_maxpacket_maxes[4] = {
 	[USB_ENDPOINT_XFER_CONTROL] = 512,
-	[USB_ENDPOINT_XFER_ISOC] = 1024,
+	[USB_ENDPOINT_XFER_ISOC] = 1544,
 	[USB_ENDPOINT_XFER_BULK] = 1024,
 	[USB_ENDPOINT_XFER_INT] = 1024,
 };
-- 
2.34.1

